src/printer.js

```diff
@@ -2977,6 +2971,50 @@ function hasLeadingOwnLineComment(text, node) {
   return res;
 }
 
+// A "leadingComment" attached to the top of the first token in the
+// return's argument, which would originally be in parens,
+// could trigger an ASI error where the parens are removed
+// and the comment bumped to the same line as the `return`,
+// while the contents of the return value are on the next line (oops!).
+//
+// This recurses the return argument, looking for the first token
+// (the leftmost leaf node) and, if it (or its parents) has any
+// leadingComments, wraps the return in parens.
+function returnArgumentHasLeadingComment(argument) {
+  const hasLeadingComment = (comments) => (
+    comments && comments.some(comment => comment.leading)
+  );
+  const getLeftSide = (node) => {
+    if (node.expressions) return node.expressions[0];
```

Comment:
So, the next phase I see with prettier is being able to plug ast transforms in-front.

It's dead simple to write an ast transform, way less so a js printer. So if we can have a ton of eslint-like plugins that work on the ast level to enforce whatever styles you like, that would be huge